desc/protoparse/parser.go

```diff
@@ -257,6 +270,46 @@ func (p Parser) ParseFilesButDoNotLink(filenames ...string) ([]*dpb.FileDescript
 	return fds, nil
 }
 
+func newImportResolver(importPaths []string) func(string) (string, error) {
+	return func(name string) (string, error) {
+		if !filepath.IsAbs(name) {
+			return name, nil
+		}
+		var currentImportPath string
+		var resolvedPath string
+		var err error
+		for _, importPath := range importPaths {
+			// If this is within this directory, we have a match.
+			if strings.HasPrefix(filepath.ToSlash(name), filepath.ToSlash(importPath)) {
+				// We have a duplicate.
+				if currentImportPath != "" {
+					return "", fmt.Errorf("%s resides in both %s and %s", name, importPath, currentImportPath)
+				}
+				currentImportPath = importPath
+				// This should always result in a relative path as we know that
+				// name is absolute, and for strings.HasPrefix to be true, import
+				// path must be absolute.
+				//
+				// We do not do strings.HasPrefix here to try to make this platform independent.
+				resolvedPath, err = filepath.Rel(importPath, name)
+				if err != nil {
+					return "", err
+				}
+			}
+		}
+		if resolvedPath == "" {
+			return "", fmt.Errorf("%s: File does not reside within any path "+
```

Comment:
I'm not sure `goprotoc` even needs to emulate this behavior. This is a weird sharp edge with `protoc` and the error even says it: it's kind of "dumb" since it doesn't do anything to try to resolve equivalent paths and only does sub-string matching. (We can do better.)